exercises/src/main/scala/fpinscala/testing/Gen.scala

```diff
@@ -13,23 +13,242 @@ The library developed in this chapter goes through several iterations. This file
 shell, which you can fill in and modify while working through the chapter.
 */
 
-trait Prop {
+package prop_trait {
+  trait Prop {
+    def check: Boolean
+
+    def &&(that: Prop): Prop =
+      new Prop { def check = Prop.this.check && that.check }
+  }
 }
 
-object Prop {
-  def forAll[A](gen: Gen[A])(f: A => Boolean): Prop = ???
+// Implementation of case class Gen - {{{
+case class Gen[+A](sample: State[RNG,A]) {
+
+  def flatMap[B](f: A => Gen[B]): Gen[B] =
+    Gen(sample.flatMap(x => f(x).sample))
+
+  def map2[B,C](g: Gen[B])(f: (A,B) => C): Gen[C] =
+    Gen(sample.map2(g.sample)(f))
+
+  def map[B](f: A => B): Gen[B] =
+    Gen(sample.map(f))
+
+  def listOfN(size: Gen[Int]): Gen[List[A]] =
+    size.flatMap(n => {
+      @scala.annotation.tailrec
+      def go(count: Int, l: List[A])(rng: RNG): (List[A], RNG) = 
+        if (count <= 0)
+          (l, rng)
+        else {
+          val (a, rng2) = sample.run(rng)
+          go(count - 1, a :: l)(rng2)
+        }
+
+      Gen(State(go(n, Nil)))
+    })
+
+  def listOfN2(size: Gen[Int]): Gen[List[A]] =
+    size.flatMap(n => Gen.listOfN2(n, this))
+
+  def unsized: SGen[A] =
+    SGen(_ => this)
+
 }
 
 object Gen {
-  def unit[A](a: => A): Gen[A] = ???
+
+  def choose(start: Int, stopExclusive: Int): Gen[Int] = {
+    require(start < stopExclusive)
```

Comment:
True. This is mainly a sanity check. I later added a similar check for the values injected by ScalaCheck.